### PR TITLE
Add ability to configure NgZone in `bootstrapApplication`

### PR DESCRIPTION
**Pull request by atscott**
see individual commits

### CODE REVIEW

File: goldens/public-api/core/index.md [hunk 1]
```diff
@@ -1049,6 +1047,12 @@ export class NgZone {
     runTask<T>(fn: (...args: any[]) => T, applyThis?: any, applyArgs?: any[], name?: string): T;
 }
 
+// @public (undocumented)
+export interface NgZoneOptions {
+    ngZoneEventCoalescing?: boolean;
```
Comment (alxhub) [inline goldens/public-api/core/index.md:1052]: Nit: I'd prefer if we didn't prefix these in the `provideZoneChangeDetection` API:
```
provideZoneChangeDetection({
  eventCoalescing: true,
  runCoalescing: false,
})
```

**Review by alxhub — APPROVED** [review]

File: packages/core/src/application_ref.ts [hunk 2]
```diff
@@ -1146,9 +1214,16 @@ export class NgZoneChangeDetectionScheduler {
   }
 }
 
-export function provideNgZoneChangeDetection(ngZone: NgZone): StaticProvider[] {
+/**
+ * Internal token used to provide verify that `provideZoneChangeDetection` is not used
```
Comment (AndrewKushnir) [inline packages/core/src/application_ref.ts:1218]:
```suggestion
 * Internal token used to verify that `provideZoneChangeDetection` is not used
```

File: packages/core/src/application_ref.ts [hunk 3]
```diff
@@ -1169,3 +1244,30 @@ export function provideNgZoneChangeDetection(ngZone: NgZone): StaticProvider[] {
     {provide: ZONE_IS_STABLE_OBSERVABLE, useFactory: isStableFactory},
   ];
 }
+
+/**
+ * Provides `NgZone`-based change detection for the application.
```
Comment (AndrewKushnir) [inline packages/core/src/application_ref.ts:1249]:
```suggestion
 * Provides `NgZone`-based change detection for the application bootstrapped using
 * `bootstrapApplication` method.
```
(may be also add a note that it's not supported in `platformBrowser().bootstrapModule`?)

File: packages/core/src/application_ref.ts [hunk 4]
```diff
@@ -445,14 +501,26 @@ export class PlatformRef {
...
       const moduleRef = createNgModuleRefWithProviders(
-          moduleFactory.moduleType, this.injector, provideNgZoneChangeDetection(ngZone));
+          moduleFactory.moduleType, this.injector,
+          internalProvideZoneChangeDetection(() => ngZone));
+
+      if ((typeof ngDevMode === 'undefined' || ngDevMode) &&
+          moduleRef.injector.get(PROVIDED_NG_ZONE, null) !== null) {
+        throw new RuntimeError(
+            RuntimeErrorCode.PROVIDER_IN_WRONG_CONTEXT,
+            '`bootstrapModule` does not support `provideZoneChangeDetection`. Use `BootstrapOptions` instead.');
```
Comment (AndrewKushnir) [inline packages/core/src/application_ref.ts:521]: 👍

**Review by AndrewKushnir — APPROVED** [review]

**Review by AndrewKushnir — APPROVED** [review]
Reviewed-for: public-api

**Comment by dylhunn** [comment]
This PR was merged into the repository by commit d7d6514add2912a18c50f190aaa8afafa313bc9e.

**Comment by dylhunn** [comment]
Hooray!

**Pull request closed without merging**